core/src/main/java/io/grpc/internal/KeepAliveManager.java

```diff
@@ -69,6 +69,7 @@ public void run() {
           // Schedule a shutdown. It fires if we don't receive the ping response within the timeout.
           shutdownFuture = scheduler.schedule(shutdown, keepAliveTimeoutInNanos,
               TimeUnit.NANOSECONDS);
+          pingFuture = null;
```

Comment:
Great catch. I updated the `transportGoesIdle` test to reproduce this "will never reschedule" bug, and then fixed it by moving the `pingFuture = null` to the top of the Runnable.